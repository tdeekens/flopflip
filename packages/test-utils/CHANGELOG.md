# @flopflip/test-utils

## 2.0.6

### Patch Changes

- Updated dependencies []:
  - @flopflip/memory-adapter@2.0.6

## 2.0.5

### Patch Changes

- [`40e212f`](https://github.com/tdeekens/flopflip/commit/40e212fba6328d6bf2f9a5f2494a4b0f6ec1b811) Thanks [@tdeekens](https://github.com/tdeekens)! - regenerate yarn.lock

- Updated dependencies [[`40e212f`](https://github.com/tdeekens/flopflip/commit/40e212fba6328d6bf2f9a5f2494a4b0f6ec1b811)]:
  - @flopflip/memory-adapter@2.0.5

## 2.0.4

### Patch Changes

- Updated dependencies [[`787b265`](https://github.com/tdeekens/flopflip/commit/787b26580f7fa973e17065c11898dada0201e9e4)]:
  - @flopflip/memory-adapter@2.0.4

## 2.0.3

### Patch Changes

- Updated dependencies [[`fb8d122`](https://github.com/tdeekens/flopflip/commit/fb8d12285062e0af4a9e66f6e5d7d1e9196a0ac2)]:
  - @flopflip/memory-adapter@2.0.3

## 2.0.2

### Patch Changes

- Updated dependencies [[`853b28d`](https://github.com/tdeekens/flopflip/commit/853b28d0d964f9cc897198d463e494881a82efcc)]:
  - @flopflip/memory-adapter@2.0.2

## 2.0.1

### Patch Changes

- [`ad5935f`](https://github.com/tdeekens/flopflip/commit/ad5935fe1d355dbd244f97f78e74fbe0e2a8a3f3) [#1320](https://github.com/tdeekens/flopflip/pull/1320) Thanks [@tdeekens](https://github.com/tdeekens)! - chore: update deps

- Updated dependencies [[`f0e8c66`](https://github.com/tdeekens/flopflip/commit/f0e8c66f008b1f6a959f463e01008844ee70d405)]:
  - @flopflip/memory-adapter@2.0.1

## 2.0.0

### Major Changes

- [`891fb29`](https://github.com/tdeekens/flopflip/commit/891fb294d5d6e016224b5a16d22760f0a55f9606) [#1287](https://github.com/tdeekens/flopflip/pull/1287) Thanks [@renovate](https://github.com/apps/renovate)! - flopflip is now built with TypeScript v4 which can cause compatibility issues if you project runs on an older version of TypeScript

## 1.1.19

### Patch Changes

- [`5598832`](https://github.com/tdeekens/flopflip/commit/5598832df710f707fe0832cd4011e26c060e24a8) [#1275](https://github.com/tdeekens/flopflip/pull/1275) Thanks [@tdeekens](https://github.com/tdeekens)! - chore: update deps

## 1.1.18

### Patch Changes

- [`eac3bd4`](https://github.com/tdeekens/flopflip/commit/eac3bd4d80fee4209ed39d7b9199916afb7f192f) [#1269](https://github.com/tdeekens/flopflip/pull/1269) Thanks [@tdeekens](https://github.com/tdeekens)! - chore: update dependencies across packages

## 1.1.17

### Patch Changes

- [`407f8e7`](https://github.com/tdeekens/flopflip/commit/407f8e7484ef25316d34f14f29b94c673ecd8aed) [#1235](https://github.com/tdeekens/flopflip/pull/1235) Thanks [@tdeekens](https://github.com/tdeekens)! - chore: update deps and migrate to TypeScript v4

* [`407f8e7`](https://github.com/tdeekens/flopflip/commit/407f8e7484ef25316d34f14f29b94c673ecd8aed) [#1235](https://github.com/tdeekens/flopflip/pull/1235) Thanks [@tdeekens](https://github.com/tdeekens)! - chore: update deps

- [`407f8e7`](https://github.com/tdeekens/flopflip/commit/407f8e7484ef25316d34f14f29b94c673ecd8aed) [#1235](https://github.com/tdeekens/flopflip/pull/1235) Thanks [@tdeekens](https://github.com/tdeekens)! - chore: update deps

## 1.1.16

### Patch Changes

- [`76354f8`](https://github.com/tdeekens/flopflip/commit/76354f8bd034b0ece14374b5eddf39858f75c7a7) [#1143](https://github.com/tdeekens/flopflip/pull/1143) Thanks [@tdeekens](https://github.com/tdeekens)! - chore: update deps

## 1.1.15

### Patch Changes

- [`32cc6a8`](https://github.com/tdeekens/flopflip/commit/32cc6a823ff9812ab2f256b69dd3f46e273feb5e) [#1102](https://github.com/tdeekens/flopflip/pull/1102) Thanks [@tdeekens](https://github.com/tdeekens)! - Update dependencies (TypeScript 3.9)

## 1.1.14

### Patch Changes

- [`ee96512`](https://github.com/tdeekens/flopflip/commit/ee96512dd32ab75e6f9790df9322d6bc27642eac) [#1089](https://github.com/tdeekens/flopflip/pull/1089) Thanks [@tdeekens](https://github.com/tdeekens)! - Updating dependencies.
