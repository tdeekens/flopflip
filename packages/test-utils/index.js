/* eslint-disable @typescript-eslint/restrict-template-expressions */
import React from 'react';
import adapter, { updateFlags } from '@flopflip/memory-adapter';
import {
  render,
  screen,
  queries,
  fireEvent,
  queryHelpers,
  buildQueries,
  act,
} from '@testing-library/react';

const mergeOptional = (defaultValue, value) =>
  value === null ? undefined : { ...defaultValue, ...value };

const defaultFlags = {
  enabledFeature: true,
  disabledFeature: false,
  variation: 'A',
};
const defaultAdapterArgs = {
  user: { key: 'nerd@tdeekens.name' },
};

const queryByAllByFlagName = (...args) =>
  queryHelpers.queryAllByAttribute('data-flag-name', ...args);
const getMultipleFlagNamesError = (c, flagName) =>
  `Found multiple elements with the 'data-flag-name' attribute of: '${flagName}'.`;
const getMissingFlagNameError = (c, flagName) =>
  `Unable to find an element with the 'data-flag-name' attribute of: '${flagName}'.`;

const [
  queryByFlagName,
  getAllByFlagName,
  getByDataFlagName,
  findAllByFlagName,
  findByFlagName,
] = buildQueries(
  queryByAllByFlagName,
  getMultipleFlagNamesError,
  getMissingFlagNameError
);

const flagNameQueries = {
  queryByFlagName,
  getAllByFlagName,
  getByDataFlagName,
  findAllByFlagName,
  findByFlagName,
};

const changeFlagVariation = (rendered, flagName, flagVariation) =>
  fireEvent.change(rendered.getByTestId('change-flag-variation'), {
    target: { value: `${flagName}:${flagVariation}` },
  });

const defaultRender = (ui, { ...rtlOptions }) => {
  const rendered = render(ui, {
    ...rtlOptions,
    queries: {
      ...queries,
      ...flagNameQueries,
    },
  });

  return rendered;
};

const fromEventString = (string) => {
  if (string === 'true') return true;
  if (string === 'false') return false;

  return string;
};

const ChangeFlagVariation = () => (
  <input
    data-testid="change-flag-variation"
    type="text"
    onChange={(event) => {
      const [flagName, flagVariation] = event.target.value.split(':');

      updateFlags({ [flagName]: fromEventString(flagVariation) });
    }}
  />
);

const renderWithAdapter = (
  ui,
  {
    components: { ConfigureFlopFlip, Wrapper = null },
    adapterArgs,
    flags,
    ...rtlOptions
  }
) => {
  const defaultedAdapterArgs = mergeOptional(defaultAdapterArgs, adapterArgs);
  const defaultedFlags = mergeOptional(defaultFlags, flags);

  const wrapUiIfNeeded = (innerElement) =>
    Wrapper ? React.cloneElement(Wrapper, null, innerElement) : innerElement;

  const rendered = render(
    wrapUiIfNeeded(
      <ConfigureFlopFlip
        adapter={adapter}
        adapterArgs={defaultedAdapterArgs}
        defaultFlags={defaultedFlags}
      >
        <>
          <ChangeFlagVariation />
          {ui}
        </>
      </ConfigureFlopFlip>
    ),
    {
      ...rtlOptions,
      queries: {
        ...queries,
        ...flagNameQueries,
      },
    }
  );
  return {
    changeFlagVariation: (flagName, flagVariation) =>
      changeFlagVariation(rendered, flagName, flagVariation),
    waitUntilConfigured: () => rendered.findByTestId('change-flag-variation'),
    ...rendered,
  };
};

const FlagsToComponent = (props) =>
  Object.entries(props.propKey ? props[props.propKey] : props).map(
    ([flagName, flagVariation]) => (
      <div key={flagName} data-flag-name={flagName}>
        {String(flagVariation)}
      </div>
    )
  );

const UntoggledComponent = (props) => (
  <span data-flag-name={props.flagName} data-flag-status="disabled">
    Feature is untoggled
  </span>
);
UntoggledComponent.displayName = 'UntoggledComponent';
UntoggledComponent.defaultProps = {
  flagName: 'isFeatureEnabled',
};
const ToggledComponent = (props) => (
  <span data-flag-name={props.flagName} data-flag-status="enabled">
    Feature is toggled
  </span>
);
ToggledComponent.defaultProps = {
  flagName: 'isFeatureEnabled',
};
ToggledComponent.displayName = 'ToggledComponent';

const components = {
  FlagsToComponent,
  UntoggledComponent,
  ToggledComponent,
};

export * from '@testing-library/react';
export {
  defaultRender as render,
  renderWithAdapter,
  adapter,
  updateFlags,
  components,
  screen,
  act,
};
