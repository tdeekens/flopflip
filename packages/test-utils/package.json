{
  "private": true,
  "name": "@flopflip/test-utils",
  "version": "3.0.23",
  "description": "Test utils for flipflop",
  "main": "index.js",
  "author": "Tobias Deekens <nerd@tdeekens.name>",
  "bugs": {
    "url": "https://github.com/tdeekens/flopflip/issues"
  },
  "homepage": "https://github.com/tdeekens/flopflip#readme",
  "keywords": [
    "feature-flags",
    "feature-toggles",
    "types"
  ],
  "dependencies": {
    "@babel/runtime": "7.16.3",
    "@flopflip/memory-adapter": "3.0.22",
    "@testing-library/jest-dom": "5.15.0",
    "@testing-library/react": "12.1.2",
    "jest-localstorage-mock": "2.4.18",
    "jest-plugin-filename": "0.0.1",
    "jest-runner-eslint": "1.0.0",
    "jest-watch-yarn-workspaces": "1.1.0",
    "react": "17.0.2"
  }
}
