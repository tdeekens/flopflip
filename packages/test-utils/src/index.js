/* eslint-disable @typescript-eslint/restrict-template-expressions */
import adapter from '@flopflip/memory-adapter';
import {
  act,
  buildQueries,
  fireEvent,
  queries,
  queryHelpers,
  render,
  screen,
} from '@testing-library/react';
import React, { cloneElement } from 'react';

const mergeOptional = (defaultValue, value) =>
  value === null ? undefined : { ...defaultValue, ...value };

const INTERNAL_FLAG_NAME = '__internalFlag__';
const INTERNAL_FLAG_VARIATION_LABEL = 'Change flag variation';
const defaultFlags = {
  [INTERNAL_FLAG_NAME]: true,
  enabledFeature: true,
  disabledFeature: false,
  variation: 'A',
};
const defaultAdapterArgs = {
  user: { key: 'nerd@tdeekens.name' },
};

const queryByAllByFlagName = (...args) =>
  queryHelpers.queryAllByAttribute('data-flag-name', ...args);
const getMultipleFlagNamesError = (c, flagName) =>
  `Found multiple elements with the 'data-flag-name' attribute of: '${flagName}'.`;
const getMissingFlagNameError = (c, flagName) =>
  `Unable to find an element with the 'data-flag-name' attribute of: '${flagName}'.`;

const [
  queryByFlagName,
  getAllByFlagName,
  getByFlagName,
  findAllByFlagName,
  findByFlagName,
] = buildQueries(
  queryByAllByFlagName,
  getMultipleFlagNamesError,
  getMissingFlagNameError
);

const flagNameQueries = {
  queryByFlagName,
  getAllByFlagName,
  getByFlagName,
  findAllByFlagName,
  findByFlagName,
};

const changeFlagVariation = (flagName, flagVariation) =>
  fireEvent.change(screen.getByLabelText('Change flag variation'), {
    target: { value: JSON.stringify({ flagName, flagVariation }) },
  });

const defaultRender = (ui, { ...rtlOptions }) => {
  // eslint-disable-next-line testing-library/render-result-naming-convention
  const rendered = render(ui, {
    ...rtlOptions,
    queries: {
      ...queries,
      ...flagNameQueries,
    },
  });

  return rendered;
};

const fromEventString = (jsonAsString) => JSON.parse(jsonAsString);

const FlagChangeField = () => {
  return (
    <>
      <label htmlFor={INTERNAL_FLAG_VARIATION_LABEL}>
        {INTERNAL_FLAG_VARIATION_LABEL}
      </label>
      <input
        id={INTERNAL_FLAG_VARIATION_LABEL}
        type="text"
        onChange={(event) => {
          const { flagName, flagVariation } = JSON.parse(event.target.value);

          adapter.updateFlags({ [flagName]: fromEventString(flagVariation) });
        }}
      />
    </>
  );
};

const FlagsToComponent = (props) => {
  return (
    <ul>
      {Object.entries(props.propKey ? props[props.propKey] : props).map(
        ([flagName, flagVariation]) => (
          <li key={flagName} data-flag-name={flagName}>
            {String(flagVariation)}
          </li>
        )
      )}
    </ul>
  );
};

const UntoggledComponent = (props) => (
  <span data-flag-name={props.flagName} data-flag-status="disabled">
    Feature is untoggled
  </span>
);
UntoggledComponent.displayName = 'UntoggledComponent';
UntoggledComponent.defaultProps = {
  flagName: 'isFeatureEnabled',
};
const ToggledComponent = (props) => (
  <span data-flag-name={props.flagName} data-flag-status="enabled">
    Feature is toggled
  </span>
);
ToggledComponent.defaultProps = {
  flagName: 'isFeatureEnabled',
};
ToggledComponent.displayName = 'ToggledComponent';

const renderWithAdapter = (
  ui,
  {
    components: { ConfigureFlopFlip, Wrapper = null },
    adapterArgs,
    flags,
    ...rtlOptions
  }
) => {
  const defaultedAdapterArgs = mergeOptional(defaultAdapterArgs, adapterArgs);
  const defaultedFlags = mergeOptional(defaultFlags, flags);

  const wrapUiIfNeeded = (innerElement) =>
    Wrapper ? cloneElement(Wrapper, null, innerElement) : innerElement;

  // eslint-disable-next-line testing-library/render-result-naming-convention
  const rendered = render(
    wrapUiIfNeeded(
      <ConfigureFlopFlip
        adapter={adapter}
        adapterArgs={defaultedAdapterArgs}
        defaultFlags={defaultedFlags}
      >
        <>
          <FlagChangeField />
          <ToggledComponent flagName={INTERNAL_FLAG_NAME} />
          {ui}
        </>
      </ConfigureFlopFlip>
    ),
    {
      ...rtlOptions,
      queries: {
        ...queries,
        ...flagNameQueries,
      },
    }
  );

  return {
    changeFlagVariation: (flagName, flagVariation) =>
      changeFlagVariation(flagName, flagVariation),
    waitUntilConfigured: async () => {
      await screen.findByLabelText(INTERNAL_FLAG_VARIATION_LABEL);
      await rendered.findByFlagName(INTERNAL_FLAG_NAME);
    },
    ...rendered,
  };
};

const components = {
  FlagsToComponent,
  UntoggledComponent,
  ToggledComponent,
};

export * from '@testing-library/react';
export {
  act,
  adapter,
  components,
  defaultRender as render,
  renderWithAdapter,
  screen,
};
