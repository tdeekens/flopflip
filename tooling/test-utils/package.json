{
  "private": true,
  "name": "@flopflip/test-utils",
  "version": "14.0.2",
  "description": "Test utils for flipflop",
  "main": "index.js",
  "author": "Tobias Deekens <nerd@tdeekens.name>",
  "bugs": {
    "url": "https://github.com/tdeekens/flopflip/issues"
  },
  "homepage": "https://github.com/tdeekens/flopflip#readme",
  "keywords": [
    "feature-flags",
    "feature-toggles",
    "types"
  ],
  "dependencies": {
    "@babel/runtime": "7.26.9",
    "@flopflip/memory-adapter": "workspace:*",
    "@flopflip/tsconfig": "workspace:*",
    "@testing-library/jest-dom": "6.6.3",
    "@testing-library/react": "16.2.0",
    "react": "18.3.1",
    "react-dom": "18.3.1"
  }
}
