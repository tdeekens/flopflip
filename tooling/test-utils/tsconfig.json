{
  "extends": "@flopflip/tsconfig/base.json",
  "exclude": ["**/dist/*", "**/vitest.config.ts"]
}
