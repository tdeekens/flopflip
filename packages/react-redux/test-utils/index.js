import { combineReducers, createStore as createReduxStore } from 'redux';

import { createFlopflipReducer, FLOPFLIP_STATE_SLICE } from '../src/';

const defaultInitialState = {};

const reducer = combineReducers({
  [FLOPFLIP_STATE_SLICE]: createFlopflipReducer(),
});
const createStore = (initialState = defaultInitialState) =>
  createReduxStore(reducer, initialState);

export { createStore };
