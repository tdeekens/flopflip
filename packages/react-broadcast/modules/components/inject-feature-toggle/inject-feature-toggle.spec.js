import React from 'react';
import { renderWithAdapter, components } from '@flopflip/test-utils';
import injectFeatureToggle from './inject-feature-toggle';
import Configure from '../configure';

const render = TestComponent =>
  renderWithAdapter(TestComponent, {
    components: { ConfigureFlopFlip: Configure },
  });

describe('without `propKey`', () => {
  describe('when feature is disabled', () => {
    const TestComponent = injectFeatureToggle('disabledFeature')(
      components.FlagsToComponent
    );

    it('should render receive the flag value as `false`', () => {
      const { queryByFlagName } = render(<TestComponent />);

      expect(queryByFlagName('isFeatureEnabled')).toHaveTextContent('false');
    });

    describe('when enabling feature', () => {
      it('should render the component representing a enabled feature', async () => {
        const { queryByFlagName, waitUntilReady, changeFlagVariation } = render(
          <TestComponent />
        );

        await waitUntilReady();

        changeFlagVariation('disabledFeature', true);

        expect(queryByFlagName('isFeatureEnabled')).toHaveTextContent('true');
      });
    });
  });

  describe('when feature is enabled', () => {
    const TestComponent = injectFeatureToggle('enabledFeature')(
      components.FlagsToComponent
    );

    it('should render receive the flag value as `true`', () => {
      const { queryByFlagName } = render(<TestComponent />);

      expect(queryByFlagName('isFeatureEnabled')).toHaveTextContent('true');
    });
  });
});

describe('with `propKey`', () => {
  describe('when feature is disabled', () => {
    const TestComponent = injectFeatureToggle(
      'disabledFeature',
      'customPropKey'
    )(components.FlagsToComponent);

    it('should render receive the flag value as `false`', () => {
      const { queryByFlagName } = render(<TestComponent />);

      expect(queryByFlagName('customPropKey')).toHaveTextContent('false');
    });
  });
});
