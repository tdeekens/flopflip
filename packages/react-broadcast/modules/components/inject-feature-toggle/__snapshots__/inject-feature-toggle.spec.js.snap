// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with \`propKey\` should match snapshot 1`] = `
<ConfigureFlopflip
  adapter={
    Object {
      "configure": [Function],
      "reconfigure": [Function],
    }
  }
  adapterArgs={Object {}}
  defaultFlags={Object {}}
  shouldDeferAdapterConfiguration={false}
>
  <ConfigureAdapter
    adapter={
      Object {
        "configure": [Function],
        "reconfigure": [Function],
      }
    }
    adapterArgs={
      Object {
        "onFlagsStateChange": [Function],
        "onStatusStateChange": [Function],
      }
    }
    defaultFlags={Object {}}
    shouldDeferAdapterConfiguration={false}
  >
    <Broadcast
      channel="@flopflip"
      compareValues={[Function]}
      value={Object {}}
    >
      <Container>
        <injectFeatureToggle(FeatureComponent)>
          <Subscriber
            channel="@flopflip"
            quiet={false}
          >
            <withProps(omitProps(FeatureComponent))
              @flopflip/flags={Object {}}
            >
              <omitProps(FeatureComponent)
                @flopflip/flags={Object {}}
                fooBar={false}
              >
                <FeatureComponent
                  fooBar={false}
                >
                  <div />
                </FeatureComponent>
              </omitProps(FeatureComponent)>
            </withProps(omitProps(FeatureComponent))>
          </Subscriber>
        </injectFeatureToggle(FeatureComponent)>
      </Container>
    </Broadcast>
  </ConfigureAdapter>
</ConfigureFlopflip>
`;

exports[`without \`propKey\` when feature is disabled should match snapshot 1`] = `
<ConfigureFlopflip
  adapter={
    Object {
      "configure": [Function],
      "reconfigure": [Function],
    }
  }
  adapterArgs={Object {}}
  defaultFlags={Object {}}
  shouldDeferAdapterConfiguration={false}
>
  <ConfigureAdapter
    adapter={
      Object {
        "configure": [Function],
        "reconfigure": [Function],
      }
    }
    adapterArgs={
      Object {
        "onFlagsStateChange": [Function],
        "onStatusStateChange": [Function],
      }
    }
    defaultFlags={Object {}}
    shouldDeferAdapterConfiguration={false}
  >
    <Broadcast
      channel="@flopflip"
      compareValues={[Function]}
      value={Object {}}
    >
      <Container>
        <injectFeatureToggle(FeatureComponent)>
          <Subscriber
            channel="@flopflip"
            quiet={false}
          >
            <withProps(omitProps(FeatureComponent))
              @flopflip/flags={Object {}}
            >
              <omitProps(FeatureComponent)
                @flopflip/flags={Object {}}
                isFeatureEnabled={false}
              >
                <FeatureComponent
                  isFeatureEnabled={false}
                >
                  <div />
                </FeatureComponent>
              </omitProps(FeatureComponent)>
            </withProps(omitProps(FeatureComponent))>
          </Subscriber>
        </injectFeatureToggle(FeatureComponent)>
      </Container>
    </Broadcast>
  </ConfigureAdapter>
</ConfigureFlopflip>
`;

exports[`without \`propKey\` when feature is disabled when enabling feature should match snapshot 1`] = `
<ConfigureFlopflip
  adapter={
    Object {
      "configure": [Function],
      "reconfigure": [Function],
    }
  }
  adapterArgs={Object {}}
  defaultFlags={Object {}}
  shouldDeferAdapterConfiguration={false}
>
  <ConfigureAdapter
    adapter={
      Object {
        "configure": [Function],
        "reconfigure": [Function],
      }
    }
    adapterArgs={
      Object {
        "onFlagsStateChange": [Function],
        "onStatusStateChange": [Function],
      }
    }
    defaultFlags={Object {}}
    shouldDeferAdapterConfiguration={false}
  >
    <Broadcast
      channel="@flopflip"
      compareValues={[Function]}
      value={
        Object {
          "flag1": true,
        }
      }
    >
      <Container>
        <injectFeatureToggle(FeatureComponent)>
          <Subscriber
            channel="@flopflip"
            quiet={false}
          >
            <withProps(omitProps(FeatureComponent))
              @flopflip/flags={
                Object {
                  "flag1": true,
                }
              }
            >
              <omitProps(FeatureComponent)
                @flopflip/flags={
                  Object {
                    "flag1": true,
                  }
                }
                isFeatureEnabled={true}
              >
                <FeatureComponent
                  isFeatureEnabled={true}
                >
                  <div />
                </FeatureComponent>
              </omitProps(FeatureComponent)>
            </withProps(omitProps(FeatureComponent))>
          </Subscriber>
        </injectFeatureToggle(FeatureComponent)>
      </Container>
    </Broadcast>
  </ConfigureAdapter>
</ConfigureFlopflip>
`;
